### PR TITLE
SafeBuffer implemented. Unsafe usage made configurable and disabled by default.

### LINKED ISSUE #5224: [TEST-FAILURE] semaphore.ClientSemaphoreTest timeout
```
06:01:01 Running com.hazelcast.client.semaphore.ClientSemaphoreTest
07:03:36 Build timed out (after 100 minutes). Marking the build as aborted.
```
https://hazelcast-l337.ci.cloudbees.com/job/Hazelcast-3.x-IbmJDK1.6/534/console

### PR DESCRIPTION
**Pull request by sancar**
SafeBuffer implemented. Unsafe usage made configurable and disabled by default.
Protocol buffer and flyweight methods are cleaned up. 
Ibm jdk 1.6 test failures related to unsafe should be fixed. 
fixes #5224 , #5219 , #5227
possible fix to #5099

### CODE REVIEW

**Comment by noctarius** [comment]
Is it a good idea to use native byteorder by default? I mean a protocol should have a defined byteorder like (normal for network protocols) Big Endian, otherwise you might end up with weird results by using different platforms for client and server (even though Big Endian as a machine type is sort of dying).

**Comment by serkan-ozal** [comment]
@sancar what is the problem with `Unsafe` at `IBM JDK 1.6` ?

**Comment by sancar** [comment]
@noctarius We are using Little Endian in protocol. There are some methods should be cleaned up that are using native byte order. They may look confusing. I will clean that up in following prs.

**Comment by sancar** [comment]
@serkan-ozal To be honest I dont exactly know yet. I will look into it more today. The problem that we see though, Unsafe seems to be returning values that its bytes all set to 1. I am suspecting there are usage in our code , that is not fit to standard. And when I test  UnsafeBuffer alone, it seem to be working correctly.

**Comment by serkan-ozal** [comment]
@sancar Ok thanks, can you send me details (via email or slack) when you have looked at today. I am just wondering about it :)

**Comment by sancar** [comment]
@david-strom-hazelcast checkstyle is failing because of https://github.com/hazelcast/hazelcast/commit/814b366a3ee91f8dfce549838e84f88fca98ee95
I will send a direct fix.

**Comment by sancar** [comment]
Checkstyle fix commit. https://github.com/hazelcast/hazelcast/commit/9dd9ab1d426ad6381755b4f4fc39374f87e070e4

**Comment by sancar** [comment]
verify

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/util/SafeBuffer.java [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+package com.hazelcast.client.impl.protocol.util;
+
+import com.hazelcast.nio.Bits;
+
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.nio.charset.Charset;
+
+public class SafeBuffer implements ClientProtocolBuffer {
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/util/SafeBuffer.java]: Can you add some javaDoc for `SafeBuffer`? Currently it is not clear why this buffer is safe and under what conditions this buffer should be used.

Also can you add some tests for `SafeBuffer`?

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/util/ClientProtocolOutput.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hazelcast.client.impl.protocol.util;
+
+/**
+ * Abstraction over a range of buffer types that allows fields to be written in client protocol.
+ * All methods use LITTLE_ENDIAN as byte order.
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/util/ClientProtocolOutput.java]: `All methods use LITTLE_ENDIAN as byte order` seems a javaDoc for an implementation.

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/util/ClientProtocolInput.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hazelcast.client.impl.protocol.util;
+
+/**
+ * Abstraction over a range of buffer types that allows fields to be read in client protocol.
+ * All methods use LITTLE_ENDIAN as byte order.
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/util/ClientProtocolInput.java]: `All methods use LITTLE_ENDIAN as byte order` seems a javaDoc for an implementation.

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/util/SafeBuffer.java (hunk 1 above)
Comment (sancar) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/util/SafeBuffer.java]: done.

**Pull request merged**